[serverless] Add EventBridge Inferred Spans

nhulston opened the pull request:
This PR will be broken into 2 separate PRs to make each PR shorter and easier to review

### What does this PR do?
- This PR fixes/adds EventBridge inferred spans for Lambda.
- Specifically, this PR supports inferred spans for `Lambda -> EventBridge -> Lambda`, `Lambda -> EventBridge -> SQS -> Lambda`, and `Lambda -> EventBridge -> SNS -> Lambda`. These are the most common use cases that customers want supported.

### Motivation
SNS and SQS are already supported. There was some code in place for EventBridge, but it didn't work properly and didn't show spans on Datadog. This PR aims to fix this, and handle other common cases like `EventBridge -> SQS` and `EventBridge -> SNS`.

### Additional Notes

- I have a [PR here](https://github.com/DataDog/dd-trace-java/pull/7613) to inject trace context into the Java tracer. At the time of writing this, the tracers for .NET and Go do not inject trace context into SQS/SNS/EventBridge yet, but that's my next project.

- The AWS API has no reliable way of getting the start time of the EventBridge span. It only gives time with precision to the second, which is not precise enough to create spans. Therefore, we try to get the start time from the tracer in the field `x-datadog-sent-timestamp`. If this field does not exist, we fall back to AWS API's start time.

- Also, the AWS API has no way of getting the EventBridge bus name, so we attempt to get this from the tracer in the field `x-datadog-bus-name` for the span resource name. If this doesn't exist, we set the resource name to "EventBridge". 

- This should not be a breaking change, but this is an area for reviewers to focus on and double check. If the user is using an outdated tracer or .NET/Go that doesn't yet support inferred spans for EB/SQS/SNS, these changes shouldn't break anything, but it should show an "EventBridge" span on any Lambda function invoked by the EventBridge event:
<img width="1336" alt="Screenshot 2024-09-16 at 4 29 37 PM" src="https://github.com/user-attachments/assets/afa0d02a-a671-4893-9009-1106f7f5dae4">

### Traces before these changes
**Lambda --> EventBridge --> Lambda**
Two different traces. Second trace is missing an EventBridge span
<img width="865" alt="Screenshot 2024-09-12 at 4 50 56 PM" src="https://github.com/user-attachments/assets/a341965c-9ff4-42dc-895d-807811bfe970">

**Lambda --> EventBridge --> SQS --> Lambda**
Missing EventBridge span
<img width="859" alt="Screenshot 2024-09-12 at 4 51 14 PM" src="https://github.com/user-attachments/assets/dea26fcc-25f8-41ce-acf5-da752b4b986b">

**Lambda --> EventBridge --> SNS --> Lambda**
Missing EventBridge span
<img width="865" alt="Screenshot 2024-09-12 at 4 51 06 PM" src="https://github.com/user-attachments/assets/f6f97b92-248d-4db6-a3f5-e935afb38e21">

### Traces after these changes
**Lambda --> EventBridge --> Lambda**
<img width="643" alt="Screenshot 2024-09-12 at 4 53 26 PM" src="https://github.com/user-attachments/assets/3d5b3c95-3807-41da-aa6d-cacba118f35a">

**Lambda --> EventBridge --> SQS --> Lambda**
<img width="643" alt="Screenshot 2024-09-12 at 4 53 18 PM" src="https://github.com/user-attachments/assets/26bf8462-3f51-449e-acf3-92cb3c210021">

**Lambda --> EventBridge --> SNS --> Lambda**
<img width="646" alt="Screenshot 2024-09-12 at 4 53 22 PM" src="https://github.com/user-attachments/assets/60506174-7e41-4052-93eb-4312e9d1888c">

### Possible Drawbacks / Trade-offs
Increased overhead

### Describe how to test/QA your changes
Run local tests with `invoke test --targets=./pkg/serverless` from the root of the project

It's more complicated to test manually. Contact me if you want a live demo of these changes. Here's how I tested:
1. Create a serverless architecture that's supported, such as a Java Lambda function that publishes an event to EventBridge, an EventBridge event bus that has a target SQS queue or SNS notifier, and another Java Lambda function that subscribes to the queue/notifier.
2. Instrument the lambda functions using `datadog-ci` 
3. Build the extension: `./scripts/publish_sandbox.sh`
4. Build the new Java tracer from [this PR](https://github.com/DataDog/dd-trace-java/pull/7613): `./gradlew publishToMavenLocal && cd ~/.m2/repository/com/datadoghq/dd-java-agent/1.40.0-SNAPSHOT && zip layer.zip dd-java-agent-1.40.0-SNAPSHOT.jar`
5. Create new layers on AWS by uploading those zip files, and replace your Lambda layers with your custom layers.
6. Set the environment variable `JAVA_TOOL_OPTIONS: -javaagent:"/opt/dd-java-agent-1.40.0-SNAPSHOT.jar"` on your lambda functions to use the new Java tracer
7. Invoke your first lambda function, and check Datadog for results. The spans should be combined into a single trace.

nhulston committed:
Add EventBridge inferred spans support

nhulston committed:
Change span resource name to the bus name

nhulston committed:
Add inferred spans support for
- Lambda --> EB --> SQS --> Lambda
- Lambda --> EB --> SNS --> Lambda

nhulston committed:
Refactoring

nhulston committed:
Refactoring

nhulston committed:
Tests

nhulston committed:
remove my debug statements

nhulston committed:
refactoring

nhulston committed:
Add release note

nhulston commented on pkg/serverless/invocationlifecycle/init.go:
```diff
@@ -139,6 +143,14 @@ func (lp *LifecycleProcessor) initFromSNSEvent(event events.SNSEvent) {
 	lp.addTag(tagFunctionTriggerEventSource, sns)
 	lp.addTag(tagFunctionTriggerEventSourceArn, trigger.ExtractSNSEventArn(event))
+
+	// Check for EventBridge event wrapped by the SNS message
+	var eventBridgeEvent events.EventBridgeEvent
+	if err := json.Unmarshal([]byte(event.Records[0].SNS.Message), &eventBridgeEvent); err == nil {
+		if eventBridgeEvent.DetailType != "" && eventBridgeEvent.Detail != nil {
+			lp.createEventBridgeInferredSpan(eventBridgeEvent)
+		}
+	}
```
This adds support for EventBridge -> SNS -> Lambda inferred spans

nhulston commented on pkg/serverless/invocationlifecycle/init.go:
```diff
@@ -150,19 +162,26 @@ func (lp *LifecycleProcessor) initFromSQSEvent(event events.SQSEvent) {
-	// test for SNS
+	// Check for SNS event wrapped by the SQS body
 	var snsEntity events.SNSEntity
-	if err := json.Unmarshal([]byte(event.Records[0].Body), &snsEntity); err != nil {
-		return
+	if err := json.Unmarshal([]byte(event.Records[0].Body), &snsEntity); err == nil {
+		if strings.ToLower(snsEntity.Type) == "notification" && snsEntity.TopicArn != "" {
+			lp.createSNSInferredSpan(snsEntity)
+			return
+		}
 	}
```
This already existing code added support for SNS -> SQS -> Lambda inferred spans

nhulston commented on pkg/serverless/trace/inferredspan/constants.go:
```diff
@@ -11,6 +11,7 @@ const (
 	apiName          = "apiname"
 	bucketARN        = "bucket_arn"
 	bucketName       = "bucketname"
+	busName          = "BusName"
```
This isn't camelCase like the other constants, but I was following the PascalCase format of `SentTimestamp`.

duncanista commented on pkg/serverless/trace/inferredspan/constants.go:
```diff
@@ -11,6 +11,7 @@ const (
 	apiName          = "apiname"
 	bucketARN        = "bucket_arn"
 	bucketName       = "bucketname"
+	busName          = "BusName"
```
this should comply with [Consistent APM span tags for AWS requests](https://docs.google.com/document/d/1450pjyhzqHTTtEvPyZGzH1cKpca7gOwtxgel_LUsR4w/edit#heading=h.w9tpabmagxvr)

nhulston replied:
Yeah I've renamed the custom fields to `x-datadog-sent-timestamp` and `x-datadog-bus-name`, and put them under `_datadog`. What do you think of this approach?

duncanista replied:
Where do those fields come from?

nhulston replied:
I just came up with them on my own. There's no existing naming convention for these fields. I tried to follow the style of the [existing Datadog headers](https://docs.datadoghq.com/tracing/trace_collection/trace_context_propagation/?tab=java#datadog-format)

duncanista commented on pkg/serverless/trace/propagation/carriers.go:
```diff
@@ -201,6 +213,51 @@ func snsEntityCarrier(event events.SNSEntity) (tracer.TextMapReader, error) {
+		return nil, errorUnsupportedPayloadType
+	}
+
+	datadogInfo, ok := detail["_datadog"].(map[string]interface{})
```
use the constant instead of `_datadog` string

nhulston replied:
Good catch

duncanista commented on pkg/serverless/trace/propagation/carriers.go:
```diff
@@ -201,6 +213,51 @@ func snsEntityCarrier(event events.SNSEntity) (tracer.TextMapReader, error) {
+		}
+	}
+
+	if carrier["x-datadog-trace-id"] == "" || carrier["x-datadog-parent-id"] == "" {
+		return nil, errorNoDDContextFound
+	}
```
Why are you checking on this specific keys?

nhulston replied:
Yeah this is kind of unnecessary. As long as we check that `_datadog` exists, we should be good

nhulston committed:
Rename `BusName` and `SentTimestamp` to
`x-datadog-bus-name` and `x-datadog-sent-timestamp`. Put them in the `_datadog` field instead.

nhulston committed:
Fix test json

nhulston committed:
Use constant `_datadog`, remove unnecessary check

nhulston committed:
Merge branch 'main' into nicholas.hulston/fix-eventbridge-spans

nhulston commented on pkg/serverless/trace/propagation/carriers.go:
```diff
@@ -201,6 +213,47 @@ func snsEntityCarrier(event events.SNSEntity) (tracer.TextMapReader, error) {
+		return nil, errorNoDDContextFound
+	}
+
+	carrier := make(tracer.TextMapCarrier)
+	for key, value := range datadogInfo {
+		if strValue, ok := value.(string); ok {
+			carrier[key] = strValue
+		}
+	}
+	return carrier, nil
```
Is there a better way to do this

duncanista replied:
@purple4reina might have a good answer for this

purple4reina replied:
Yes. Here's another option: https://github.com/DataDog/datadog-agent/compare/main...rey.abolofia%2Fnhulston

Notice one thing about this code. I rely on go typing and json unmarshalling to do a lot of the work for me. I change the `Detail` field on the `EventBridgeEvent` to a `map[string]map[string]string`. Since that's the type you're wanting, just force it into that from the start. It's also just fun to say, so a little bonus benefit there too 😉

nhulston commented on pkg/serverless/trace/propagation/carriers.go:
```diff
@@ -201,6 +213,47 @@ func snsEntityCarrier(event events.SNSEntity) (tracer.TextMapReader, error) {
+		return nil, errorNoDDContextFound
+	}
+
+	carrier := make(tracer.TextMapCarrier)
+	for key, value := range traceContext {
+		if strValue, ok := value.(string); ok {
+			carrier[key] = strValue
+		}
+	}
+
+	return carrier, nil
```
and this

purple4reina commented on pkg/serverless/trace/inferredspan/span_enrichment.go:
```diff
@@ -288,17 +288,29 @@ func (inferredSpan *InferredSpan) EnrichInferredSpanWithSQSEvent(eventPayload ev
 // payload to enrich the current inferred span. It applies a
 // specific set of data to the span expected from an EventBridge event.
 func (inferredSpan *InferredSpan) EnrichInferredSpanWithEventBridgeEvent(eventPayload events.EventBridgeEvent) {
```
Am I understanding this correctly? This `EnrichInferredSpanWithEventBridgeEvent` method existed in the extension already, but it just wasn't being used anywhere?

nhulston replied:
Yes that's correct. It was supposed to be called [here](https://github.com/DataDog/datadog-agent/blob/3b96ef1a258e773bab62a1d0b355bb6186a9e51b/pkg/serverless/invocationlifecycle/init.go#L109-L111)

purple4reina commented on pkg/serverless/trigger/events/events.go:
```diff
@@ -228,9 +228,10 @@ type KinesisRecord struct {
+	DetailType string                 `json:"detail-type"`
+	Source     string                 `json:"source"`
+	Time       time.Time              `json:"time"` // Time only has precision to the nearest second
+	Detail     map[string]interface{} `json:"detail"`
```
You shouldn't need to add the `json:...` for source, time, and detail. Since we're only ever unmarshalling the json, go will ignore the casing. See https://go.dev/play/p/qcrfqjCfZ7J
